numpy/f2py/cfuncs.py

```diff
@@ -621,71 +630,118 @@
 }"""
 needs['try_pyarr_from_string'] = ['STRINGCOPYN', 'PRINTPYOBJERR', 'string']
 cfuncs['try_pyarr_from_string'] = """\
-static int try_pyarr_from_string(PyObject *obj,const string str) {
+/*
+  try_pyarr_from_string copies str[:len(obj)] to a mutable Python
+  string-like obj.
+
+  if the specified len==-1, str must be null-terminated.
+*/
+static int try_pyarr_from_string(PyObject *obj,
+                                 const string str, const int len) {
+#ifdef DEBUGCFUNCS
+fprintf(stderr,\"try_pyarr_from_string(str='%s',
+               len=%d, obj=%p)\\n\",(char*)str,len, obj);
+#endif
     PyArrayObject *arr = NULL;
-    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL)))
-        { STRINGCOPYN(PyArray_DATA(arr),str,PyArray_NBYTES(arr)); }
-    return 1;
+    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL))) {
+        string buf = PyArray_DATA(arr);
+        int n = len;
```

Comment:
Is `int` the correct type here? Py_ssize_t is used for python unicode/bytes strings.

Reply:
The code here and Python unicode/bytes strings have no direct relation. In fact, using `npy_intp` instead of `Py_ssize_t` would make more sense. Although, I think it is impossible to achieve complete uniformity of the types for handling the buffer lengths. For instance, the C function `strlen` returns `size_t`.

That said, I agree that `int` should be replaced with `npy_intp` (f2py needs to follow NumPy array API more than of Python unicode API) but not in this PR because the replacement should be done everywhere in the f2py codebase at once.

Reply:
Now using `npy_intp` instead of `int`.